warp10/src/main/java/io/warp10/script/functions/DURATIONBUCKETIZE.java

```diff
@@ -0,0 +1,370 @@
+          }
+          series.add((GeoTimeSerie) o);
+        }
+      } else {
+        throw new WarpScriptException(getName() + " expects a Geo Time Series or a list of Geo Time Series as first parameter.");
+      }
+    }
+
+    if (!(params.get(params.size() - 4) instanceof WarpScriptBucketizerFunction) && !(params.get(params.size() - 4) instanceof Macro) && null != params.get(params.size() - 4)) {
+      throw new WarpScriptException(getName() + " expects a bucketizer function or a macro as fourth to last parameter.");
+    }
+
+    if (!(params.get(params.size() - 3) instanceof Long) || !(params.get(params.size() - 2) instanceof String) || !(params.get(params.size() - 1) instanceof Long)) {
+      throw new WarpScriptException(getName() + " expects lastbucket, bucketduration, bucketcount (and optionally timezone) as last parameters.");
+    }
+
+    Object bucketizer = params.get(params.size() - 4);
+    long lastbucket = (long) params.get(params.size() - 3);
+    String bucketduration = (String) params.get(params.size() - 2);
+    long bucketcount = (long) params.get(params.size() - 1);
+
+    //
+    // Check that lastbucket is not 0
+    //
+
+    if (0 == lastbucket) {
+      throw new WarpScriptException(getName() + " does not allow lastbucket to be 0. It must be specified.");
+    }
+
+    //
+    // Check that bucketcount is not negative or null and not over maxbuckets
+    //
+
+    if (bucketcount < 0) {
+      throw new WarpScriptException(getName() + " expects a positive bucketcount.");
+    }
+
+    long maxbuckets = (long) stack.getAttribute(WarpScriptStack.ATTRIBUTE_MAX_BUCKETS);
+    if (bucketcount > maxbuckets) {
+      throw new WarpScriptException(getName() + " error: bucket count (" + bucketcount + ") would exceed maximum value of " + maxbuckets);
+    }
+
+    //
+    // Check that input gts are not already duration-bucketized
+    //
+
+    for (GeoTimeSerie gts : series) {
+      Map<String, String> attributes = gts.getMetadata().getAttributes();
+      if (attributes.get(DURATION_ATTRIBUTE_KEY) != null || attributes.get(OFFSET_ATTRIBUTE_KEY) != null || attributes.get(TIMEZONE_ATTRIBUTE_KEY) != null) {
+        throw new WarpScriptException(getName() + " expects GTS for which the attributes " + DURATION_ATTRIBUTE_KEY + ", " + OFFSET_ATTRIBUTE_KEY + " and " + TIMEZONE_ATTRIBUTE_KEY + " are not set. If an input GTS is supposed to be already duration-bucketized, duration-unbucketize it first before applying a new duration-bucketization.");
+      }
+    }
+
+    //
+    // Check nullity of bucketizer
+    //
+
+    if (null == bucketizer) {
+      throw new WarpScriptException(getName() + " expects a non null bucketizer.");
+    }
+
+    //
+    // Convert duration to joda.time.Period
+    //
+
+    if (!DURATION_RE.reset(bucketduration).matches()) {
+      throw new WarpScriptException(getName() + " expects the bucketduration parameter to be a valid ISO8601 duration with positive coefficients.");
```

Comment:
Why positive coefficients? Wouldn't P1M-1d be legitimate to support?

Reply:
Yes it could be. We could just check that they the overall duration is positive.

Reply:
Actually this one is tricky, the duration can be positive from some date and negative from others. For example, allowing P1M-31DT1H would make the computation go back and forth.

Reply:
You can throw an exception if you encounter an inconsistency while generating the indices